### PR TITLE
[FileCheck] Don't use regex to find prefixes

### PR DESCRIPTION
**Pull request by nikic**
FileCheck currently compiles a regular expression of the form `Prefix1|Prefix2|...` and uses it to find the next prefix in the input.

If we had a fast regex implementation, this would be a useful thing to do, as the regex implementation would be able to match multiple prefixes more efficiently than a naive approach. However, with our actual regex implementation, finding the prefixes basically becomes O(InputLen * RegexLen * LargeConstantFactor), which is a lot worse than a simple string search.

Replace the regex with StringRef::find(), and keeping track of the next position of each prefix. There are various ways this could be improved on, but it's already significantly faster that the previous approach.

For me, this improves check-llvm time from 138.5s to 132.5s, so by around 4-5%.

For vector-interleaved-load-i16-stride-7.ll in particular, test time drops from 5s to 2.5s.

### CODE REVIEW

**Comment by jayfoad** [comment]
Nice! `test/CodeGen/AMDGPU/memory-legalizer-*` may benefit from this.

**Comment by nikic** [comment]
> Nice! `test/CodeGen/AMDGPU/memory-legalizer-*` may benefit from this.

I checked, and it doesn't. These tests spend most time in `llc` rather than `FileCheck`. More specifically, they spend most of the time in `-verify-machineinstrs`. `verifyUseLists()` and `getMinimalPhysRegClassLLT()` together take up about 2/3 of the profile.

File: llvm/lib/FileCheck/FileCheck.cpp [hunk 1]
```diff
@@ -1634,6 +1634,59 @@ static size_t SkipWord(StringRef Str, size_t Loc) {
...
+    for (StringRef Prefix : CommentPrefixes)
+      Prefixes.push_back({Prefix, Input.find(Prefix)});
+
+    // Sort by descending length.
+    llvm::sort(Prefixes,
+               [](auto A, auto B) { return A.first.size() > B.first.size(); });
+  }
+
+  /// Find the next match of a prefix in Buffer.
+  std::optional<StringRef> match(StringRef Buffer) {
```
Comment (arsenm) [inline llvm/lib/FileCheck/FileCheck.cpp:1670]: Don't see the point of the optional, the empty StringRef will naturally act like none

**Review by tstellar — APPROVED** [review]
LGTM.

**Pull request merged**